src/ioctls/sev.rs

```diff
@@ -0,0 +1,680 @@
+pub struct LaunchUpdate(Handle);
+pub struct LaunchSecret(Handle, Measurement);
+pub struct Running(Handle, Measurement);
+#[allow(dead_code)]
+pub struct SendUpdate(Handle, Measurement);
+#[allow(dead_code)]
+pub struct ReceiveUpdate(Handle, Measurement);
+
+/// Wrapper over `/dev/sev` Fd
+pub struct SevFd {
+    sev: File,
+}
+
+impl AsRawFd for SevFd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.sev.as_raw_fd()
+    }
+}
+
+/// Helper function to create a new `SevFd`.
+///
+/// This should not be exported as a public function because the preferred way is to use
+/// `new_with_vm` from `SevLaunch`.
+fn new_sevfd(sev: File) -> SevFd {
+    SevFd { sev }
+}
+
+/// KVM SEV launch contex, using a `state machine` pattern.
+pub struct SevLaunch<S> {
+    pub sev: SevFd,
+    pub vm: VmFd,
+    pub state: S,
+}
+
+impl<T> SevLaunch<T> {
+    /// Wrapper over `KVM_MEMORY_ENCRYPT_OP`.
+    ///
+    /// See documentation for `KVM_MEMORY_ENCRYPT_OP`.
+    ///
+    /// # Arguments
+    /// * `cmd` - command argument to `KVM_MEM_ENCRYPT_OP` (in). For details check the `kvm_sev_cmd`
+    ///           structure in the
+    ///           [KVM SEV API doc](https://www.kernel.org/doc/html/latest/virt/kvm/amd-memory-encryption.html).
+    /// * `data` - struct containing arguments specific to command (out).
+    ///
+    /// Returns data or an error otherwise.
+    fn sev_ioctl<U>(&self, cmd: SevCmd, mut data: U) -> Result<U> {
+        let mut input = kvm_sev_cmd {
+            id: cmd as u32,
+            data: &mut data as *mut _ as u64,
+            error: 0,
+            sev_fd: self.sev.as_raw_fd() as u32,
+        };
+
+        // The ioctl is safe because we allocated the struct and we know the
+        // kernel will write exactly the size of the struct.
+        let ret = unsafe { ioctl_with_mut_ref(&self.vm, KVM_MEMORY_ENCRYPT_OP(), &mut input) };
+        if ret == 0 {
+            Ok(data)
+        } else {
+            Err(errno::Error::last())
+        }
+    }
+}
+
+impl SevLaunch<Uninit> {
+    /// Instantiate a new SEV launch context with VmFd.
+    pub fn new_with_vm(vm: VmFd) -> Result<Self> {
+        let sev = OpenOptions::new().read(true).write(true).open("/dev/sev")?;
+
+        // Safe because we are the owners of the fd.
+        Ok(SevLaunch {
+            sev: new_sevfd(sev),
+            vm: vm,
+            state: Uninit,
+        })
+    }
+
+    /// Initialize the SEV platform context.
+    /// In a typical workflow, this command should be the first command issued.
+    pub fn init(self) -> Result<SevLaunch<Init>> {
+        self.sev_ioctl(SevCmd::Init, ())?;
+
+        Ok(SevLaunch {
+            sev: self.sev,
+            vm: self.vm,
+            state: Init,
+        })
+    }
+}
+
+impl From<SevLaunch<Uninit>> for SevLaunch<Init> {
+    fn from(launch: SevLaunch<Uninit>) -> SevLaunch<Init> {
+        launch
+            .init()
+            .expect("Launch state transition failed from Uninit to Init")
```

Comment:
I'm not familiar with how things are supposed to work on AMD SEV. If the transition fails, is it considered a serious enough program error that it terminates the VMM? Or is it recoverable and we can keep the VMM alive, turn this into a `TryFrom`, and propagate the error?